packages/dashboard/src/components/grid/gestureable.tsx

```diff
@@ -0,0 +1,248 @@
+import type { PointerEventHandler, PropsWithChildren } from 'react';
+import React, { useEffect, useState } from 'react';
+import { useDrag, useDrop } from 'react-dnd';
+import { useKeyPress } from '~/hooks/useKeyPress';
+import type { Position } from '~/types';
+import { MouseClick } from '~/types';
+import { ItemTypes } from '../dragLayer/itemTypes';
+import { gestureable } from '../internalDashboard/gestures/determineTargetGestures';
+import { DASHBOARD_CONTAINER_ID, getDashboardPosition } from './getDashboardPosition';
+import type { DashboardState } from '~/store/state';
+import type { ComponentPaletteDraggable } from '../palette/types';
+import { SizedGrid } from './sizedGrid';
+
+export type DragEvent = {
+  target?: EventTarget;
+  start: Position;
+  end: Position;
+  vector: Position;
+  union: boolean;
+};
+export type PointClickEvent = {
+  target?: EventTarget;
+  position: Position;
+  union: boolean;
+};
+
+export type DropEvent = {
+  position: Position;
+  item: {
+    componentTag: string;
+  };
+};
+
+export type GesturableGridProps = PropsWithChildren<{
+  readOnly: boolean;
+  grid: DashboardState['grid'];
+  click: (e: PointClickEvent) => void;
+  dragStart: (e: DragEvent) => void;
+  drag: (e: DragEvent) => void;
+  dragEnd: (e: DragEvent) => void;
+  drop: (e: DropEvent) => void;
+}>;
+
+const trackPosition = (defaultPosition: Position) => {
+  let position = defaultPosition;
+
+  return {
+    getPosition: () => position,
+    setPosition: (newPosition: Position) => {
+      position = newPosition;
+    },
+  };
+};
+
+const defaultDelta = { x: 0, y: 0 };
+
+const deltaTracker = trackPosition(defaultDelta);
+const startTracker = trackPosition(defaultDelta);
+const endTracker = trackPosition(defaultDelta);
+
+const constrainPosition = (params: {
+  position: Position;
+  gridSize: { width: number; height: number; x: number; y: number };
+}) => {
+  const { position, gridSize } = params;
+  return {
+    x: Math.min(Math.max(position.x, 0), gridSize.width),
+    y: Math.min(Math.max(position.y, 0), gridSize.height),
+  };
+};
+
+export const GestureableGrid: React.FC<GesturableGridProps> = ({
```

Comment:
There's a huge number of hooks (state and side-effects) in this component. It's very difficult for me to wrap my head around what's going on. Could you refactor this with custom hooks and add comments where beneficial?